File: android/app/src/main/java/com/pauseforpaws/app/CorridorAlertPlugin.java
package com.pauseforpaws.app;

import android.Manifest;
import android.content.Intent;
import android.os.Build;
import com.getcapacitor.JSObject;
import com.getcapacitor.PermissionState;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.Permission;
import com.getcapacitor.annotation.PermissionCallback;

// Bridges the JS UI to the native background corridor-alert foreground service.
@CapacitorPlugin(
    name = "CorridorAlert",
    permissions = {
        @Permission(strings = { Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION }, alias = "location"),
        @Permission(strings = { Manifest.permission.POST_NOTIFICATIONS }, alias = "notifications")
    }
)
public class CorridorAlertPlugin extends Plugin {

    @PluginMethod
    public void start(PluginCall call) {
        if (getPermissionState("location") != PermissionState.GRANTED) {
            requestPermissionForAlias("location", call, "locationCallback");
            return;
        }
        if (Build.VERSION.SDK_INT >= 33 && getPermissionState("notifications") != PermissionState.GRANTED) {
            requestPermissionForAlias("notifications", call, "notificationCallback");
            return;
        }
        startService(call);
    }

    @PermissionCallback
    private void locationCallback(PluginCall call) {
        if (getPermissionState("location") == PermissionState.GRANTED) {
            start(call);
        } else {
            call.reject("Location permission is required for corridor alerts.");
        }
    }

    @PermissionCallback
    private void notificationCallback(PluginCall call) {
        start(call);
    }

    private void startService(PluginCall call) {
        Intent intent = new Intent(getContext(), CorridorAlertService.class);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            getContext().startForegroundService(intent);
        } else {
            getContext().startService(intent);
        }
        JSObject result = new JSObject();
        result.put("running", true);
        call.resolve(result);
    }

    @PluginMethod
    public void stop(PluginCall call) {
        Intent intent = new Intent(getContext(), CorridorAlertService.class);
        getContext().stopService(intent);
        JSObject result = new JSObject();
        result.put("running", false);
        call.resolve(result);
    }
}

File: ios/App/App/CorridorAlertPlugin.swift
import Foundation
import Capacitor

// Bridges the JS UI to the native background corridor-alert monitor (mirrors the Android plugin).
@objc(CorridorAlertPlugin)
public class CorridorAlertPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "CorridorAlertPlugin"
    public let jsName = "CorridorAlert"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "start", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stop", returnType: CAPPluginReturnPromise)
    ]

    @objc func start(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            CorridorAlertMonitor.shared.start { error in
                if let error = error {
                    call.reject(error.localizedDescription)
                } else {
                    call.resolve(["running": true])
                }
            }
        }
    }

    @objc func stop(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            CorridorAlertMonitor.shared.stop()
            call.resolve(["running": false])
        }
    }
}
